### PR TITLE
[modified] modified mlops-plugin-parser for parsing updated json format.

### PR DESCRIPTION
**Pull request by yeonykim2**
- json file is updated from separated files to a single file.
 - add parsing updated json format code in ``mlops-plugin-parser.cc``.
 - an example of the json file is shown below. The name of the json file is ``rpk_config.json``.

```
{  
    "models": [  
        {  
            "name": "image_classification",  
            "model": "mobilenet_v2_1.0_224.tflite",  
            "activate": true,
            "description": "Image classification model for mobile devices (version 2.1.0)"  
	},
        {  
            "name": "object_detection",  
            "model": "ssd_mobilenet_v1_100_320_uint8_default_1.tflite",
            "activate": true,  
            "description": "Object detection model for system (version 1)"  
        }  
    ],  
    "pipelines": [
        {  
            "name": "pipeline1",  
            "pipeline": "videotestsrc is-live=true ! videoconvert ! tensor_converter ! tensor_sink async=false"
        },  
        {  
            "name": "pipeline2",  
            "pipeline": "audiotestsrc ! audioconvert ! audio/x-raw,format=S16LE,rate=16000 ! tensor_converter ! tensor_sink"
        }  
    ],  
    "resources": [  
        {  
            "name": "ml-resource",  
            "path": [
                "orange.png",
                "orange.raw"
            ],
            "description": "The is sample data file." 
        },  
        {  
            "name": "ml-resource2",  
            "path": "labels.txt",  
            "description": "This is another sample data file."  
        }  
    ]  
}  

```

### CODE REVIEW

File: plugin-parser/mlops-plugin-parser.cc [hunk 1]
```diff
@@ -202,25 +161,38 @@ _parse_json (const gchar *json_path, mlsvc_json_type_e json_type, const gchar *a
...
           const gchar *desc = json_object_get_string_member (object, "description");
-          const gchar *clear = json_object_get_string_member (object, "clear");
+          JsonNode *path_node = json_object_get_member (object, "path");
+          JsonArray *path_array = NULL;
+          guint path_len;
 
-          if (!name || !path) {
-            _E ("Failed to get name or path from json file '%s'.", json_file);
+          if (!name) {
+            _E ("Failed to get name or path from MLSVC_JSON_RESOURCE.");
```
Comment (jaeyun-jung) [inline plugin-parser/mlops-plugin-parser.cc:170]: plz fix log message 
```
if !name
  "Failed to get name from MLSVC_JSON_RESOURCE."

if path_len == 0
  "Failed to get path from MLSVC_JSON_RESOURCE."
```

File: plugin-parser/mlops-plugin-parser.cc [hunk 2]
```diff
@@ -202,25 +161,38 @@ _parse_json (const gchar *json_path, mlsvc_json_type_e json_type, const gchar *a
...
+          if (path_len == 0U) {
+            _E ("Failed to get path array from resource json.");
+            continue;
           }
 
-          ret = ml_agent_resource_add (
+          for (guint pidx = 0; pidx < path_len; pidx++) {
+            const gchar *path;
+
+            if(path_array)
```
Comment (jaeyun-jung) [inline plugin-parser/mlops-plugin-parser.cc:188]: indent rule `if (`

File: plugin-parser/mlops-plugin-parser.cc [hunk 3]
```diff
@@ -238,6 +210,73 @@ _parse_json (const gchar *json_path, mlsvc_json_type_e json_type, const gchar *a
...
+  }
+
+  object = json_node_get_object (root);
+  members = json_object_get_members(object);
+
+  for (iter = members; iter !=NULL ; iter = iter->next) {
+    const gchar *name = (const gchar*)iter->data;
+    JsonNode *node;
+
+    if (g_ascii_strcasecmp (name, "models") == 0) {
```
Comment (jaeyun-jung) [inline plugin-parser/mlops-plugin-parser.cc:259]: json node name:
"model" "pipeline" "resource" vs "models" "pipelines" "resources" 

or 

if (g_ascii_strcasecmp (name, "model") == 0 || g_ascii_strcasecmp (name, "models") == 0) {

File: plugin-parser/mlops-plugin-parser.cc (hunk 2 above)
Comment (jaeyun-jung) [reply plugin-parser/mlops-plugin-parser.cc:188]: this parser depends on tizen feature and cannot check the indent rule in ci.
I recommend using clang-format (e.g., clang-format-12 -i --style=file ./plugin-parser/mlops-plugin-parser.cc) on your host pc.

File: plugin-parser/mlops-plugin-parser.cc (hunk 1 above)
Comment (yeonykim2) [reply plugin-parser/mlops-plugin-parser.cc]: thanks, updated it!

File: plugin-parser/mlops-plugin-parser.cc (hunk 2 above)
Comment (yeonykim2) [reply plugin-parser/mlops-plugin-parser.cc]: thanks, updated it!

File: plugin-parser/mlops-plugin-parser.cc (hunk 3 above)
Comment (yeonykim2) [reply plugin-parser/mlops-plugin-parser.cc]: How about using "models", "pipelines", and "resources" as the default or recommend?
"models" json node has "model" value in sample json.

But, i modified the code that "model" "pipeline" "resource" are also possible!

**Review by jaeyun-jung — APPROVED** [review]

File: plugin-parser/mlops-plugin-parser.cc [hunk 4]
```diff
@@ -202,25 +161,38 @@ _parse_json (const gchar *json_path, mlsvc_json_type_e json_type, const gchar *a
...
 
-          ret = ml_agent_resource_add (
-              name, path, desc ? desc : "", app_info ? app_info : "");
+          for (guint pidx = 0; pidx < path_len; pidx++) {
+            const gchar *path;
+
+            if (path_array)
+              path = json_array_get_string_element (path_array, pidx);
+            else
+              path = g_strdup (json_node_get_string (path_node));
```
Comment (jaeyun-jung) [inline plugin-parser/mlops-plugin-parser.cc:191]: fix mem leak case here,
path = json_node_get_string (path_node);

File: plugin-parser/mlops-plugin-parser.cc [hunk 5]
```diff
@@ -202,25 +161,38 @@ _parse_json (const gchar *json_path, mlsvc_json_type_e json_type, const gchar *a
...
+
+            ret = ml_agent_resource_add (
+                name, path, desc ? desc : "", app_info ? app_info : "");
+          }
```
Comment (jaeyun-jung) [inline plugin-parser/mlops-plugin-parser.cc:195]: move line 197~200 in the loop, need to print log message for each array item.

File: plugin-parser/mlops-plugin-parser.cc [hunk 6]
```diff
@@ -238,6 +216,73 @@ _parse_json (const gchar *json_path, mlsvc_json_type_e json_type, const gchar *a
 /**
  * @brief Handle INSTALL process. Tizen app-installer invoke this function.
  */
+static void
```
Comment (niley7464) [inline plugin-parser/mlops-plugin-parser.cc:219]: I think Line 216~218 is for Tizen app-installer using the Tizen appfw below.
https://review.tizen.org/git/?p=platform/core/appfw/pkgmgr-info.git;a=summary

How about changing comment liked this?

```
/**
 * @brief Get config from json to app_info
 */
```

**Review by niley7464 — APPROVED** [review]
LGTM 😄

Now we can use single `rpk_config.json` file to config!

I'm not asking you to add it to this PR, but it would be nice to add a unit test using the example json [file](https://github.com/nnstreamer/deviceMLOps.MLAgent/pull/37#issue-2199310804) you provided.

Also I left some minor comments about Tizen appfw related.
If you have time, I would appreciate it if you could update your comments.

File: plugin-parser/mlops-plugin-parser.cc (hunk 6 above)
Comment (yeonykim2) [reply plugin-parser/mlops-plugin-parser.cc:219]: Thanks reviews.
I'm going to add the example about the json unit test. 
and i fixed @brief.

File: plugin-parser/mlops-plugin-parser.cc [hunk 7]
```diff
@@ -202,30 +161,49 @@ _parse_json (const gchar *json_path, mlsvc_json_type_e json_type, const gchar *a
...
-            ml_agent_resource_delete (name);
+          if (path_len == 0U) {
+            _E ("Failed to get path from MLSVC_JSON_RESOURCE.");
+            continue;
           }
 
-          ret = ml_agent_resource_add (
-              name, path, desc ? desc : "", app_info ? app_info : "");
+          for (guint pidx = 0; pidx < path_len; pidx++) {
+            g_autofree const gchar *path = NULL;
```
Comment (jaeyun-jung) [inline plugin-parser/mlops-plugin-parser.cc:186]: param path is const char* from json, this should not be freed.
you can call json-glib function here.
```
const gchar *path = NULL;

if (path_array)
  path = json_array_get_string_element (path_array, pidx);
else
  path = json_node_get_string (path_node);
```

Comment (yeonykim2) [reply plugin-parser/mlops-plugin-parser.cc]: thanks, updated it!

**Comment by jaeyun-jung** [comment]
@anyj0527 @yeonykim2 Now merging this PR, please prepare next commit for rpk installation.

**Pull request merged**